File: 0_9_topologicalSort.java
import java.util.*;
public class TopologicalSort {
    // number of vertices
    private int vertices;
    //  adjacency list
    private List<Integer> adjList[];
    // visit 
    private boolean visit[];
    // post
    private List<Integer> post = new ArrayList<>();

    @SuppressWarnings("unchecked") TopologicalSort(int vertex){
        vertices = vertex;
        adjList = new ArrayList[vertex];
        visit = new boolean[vertex];
        for (int x=0; x< vertex; x++){
            adjList[x] = new ArrayList<Integer>();
            visit[x] = false;
        }
    }

    //  add edge
    void addEdge(int vertex1, int vertex2){
        adjList[vertex1].add(vertex2);
    }

    // postVisit
    void postVisit(int vertex){
        post.add(vertex);
    }

    // explore
    void Explore(int vertex){
        //  mark as bvisited
        visit[vertex]=true;

        //iterate
        Iterator<Integer> i = adjList[vertex].listIterator();
        while(i.hasNext()){
            int ver = i.next();
            if (visit[ver]==false){
                Explore(ver);
            }
        }
        postVisit(vertex);
    }

    //dfs
    void dfs(){
        for (int vertex=0; vertex<vertices; vertex++){
            if (visit[vertex]==false){
                Explore(vertex);
            }
        }
    }


    public static void main(String[] args){
        Scanner scanner =new Scanner(System.in);

        int vertex = scanner.nextInt();
        int edge = scanner.nextInt();

        TopologicalSort graph = new TopologicalSort(vertex);

        for (int x=0; x<edge; x++){
            int vertex1 = scanner.nextInt();
            int vertex2 = scanner.nextInt();

            //  add to the adjacent list
            graph.addEdge(vertex1, vertex2);
        }
        scanner.close();

        graph.dfs();
        
        Collections.reverse(graph.post);
        System.out.print(graph.post);
    }
}

File: 0_9_topologicalSort.py
# directed a cyclc graph
import threading

class Graph:
    def __init__(self, vertex):
        # number of vertices
        self.vertices = vertex
        # adjacency list
        self.adjList = [[] for x in range(vertex)]
        # postOrder
        self.post = []
        # visit
        self.visit = [False for x in range(vertex)]

    # add edge
    def addEdge(self, vertex1, vertex2):
        self.adjList[vertex1].append(vertex2)

    # postVisit
    def postVisit(self, vertex):
        self.post.append(vertex)

    #  explore
    def Explore(self, vertex):
        #  first mark as visited 
        self.visit[vertex]=True

        # now check iyt's all neighbour
        for ver in self.adjList[vertex]:
            # now check visited or not
            if self.visit[ver]==False:
                self.Explore(ver)

        #  now assign this vertex in the postVisit
        self.postVisit(vertex)

    # dfs
    def dfs(self):
        for x in range(self.vertices):
            if self.visit[x]==False:
                self.Explore(x)


def main():
    vertex, edge = [int(x) for x in input().split()]

    graph = Graph(vertex)

    #  assign all the edges in the graph
    for x in range(edge):
        vertex1, vertex2 = [int(y) for y in input().split()]

        graph.addEdge(vertex1, vertex2)


    # do dfs
    graph.dfs()

    #  now reverse of the post is the topologivcal order
    graph.post.reverse()
    print(*graph.post)

threading.Thread(target=main).start()
'''
testcase

input
6 6
5 2
5 0
4 0
4 1
2 3
3 1
output
5 4 2 3 1 0
'''
